### DISCUSSION TITLE
Character step up

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#189] [Q&A] [answered]
Hmm, I am having some issues with setting the step height of the character during walk step. The example has a few stairs to walk on and sets the height of the step the character can step on here:

https://github.com/jrouwe/JoltPhysics.js/blob/e2b946880fa893cad08bb81fba4a95cf3daec5de/Examples/character_virtual.html#L181-L182

This makes it so that some steps can be walked on, and others not, presumably because they are too high (`vec.y = 0.4`). However, if I hardcode the step height to something higher, the character still cannot walk on those:

```js
// const vec = characterUp.clone().multiplyScalar(updateSettings.mWalkStairsStepUp.Length());
// updateSettings.mWalkStairsStepUp.Set(vec.x, vec.y, vec.z);
updateSettings.mWalkStairsStepUp.Set(0, 1, 0);
```

In my application the behavior is a little different - it kind of works, but not precisely, e.g. if I set max height step to be 40cm, it can still climb to 50cm, but not 60cm. Or if I set height to be 10cm, it can climb 20cm, but not 30cm. Is the height measured from the character feet?

Edit:
Also, adjusting `mStickToFloorStepDown/mWalkStairsStepDownExtra` don't seem to affect the character. No matter how large the translation is set, it falls from the high steps by gravity, without sticking to the steps.

**Comment by PhoenixIllusion** [comment, marked as answer]
Do you have an example?

The height should be measured from the base according to the debug-render, but if you are using a capsule then it is important to realize that the curved bottom of the capsule may be able to catch the ledge of the step and pull it up, so a capsule testing with a 0.1 walk-step upward may be able to hook the ledge of a 0.2+ height step and make the jump.

![image](https://github.com/user-attachments/assets/65120e66-9ac8-408c-9016-a9fbd2418478)


If you add the following to your own debug build, you can enable the DebugRender of the Character Virtual and see how it attempts to sweep up and forward with the shape to make the next step (assuming it is a potentially viable step). There is even a little white ray that points up to indicate how high it can make the next step, and you can sometimes see it render your test wireframe at that elevation to attempt to make it jump the step up.

```cpp
///inside Jolt Interface constructor
CharacterVirtual::sDrawConstraints = true;
CharacterVirtual::sDrawWalkStairs = true;
CharacterVirtual::sDrawStickToFloor = true;
```

To fully debug the walking up steps, it helps to use a cylinder or cube shape so that the bottom contact is fully flat and doesn't risk hooking an edge that it shouldn't.

Your end value may be off by 0.02 or so due to the registered padding I believe, but you can test via a box character, setup a row of stairs that increments by 0.01 and use that to test the actual height of the walk-up vs actual experienced.

I ran between 0.73 and 0.78 high stairs, and a registered 0.75 upward vector, and hit could walk-step over 0.76 but not 0.77 due to the padding I think.

**Reply by jrouwe** [reply]
This is correct. The only thing I have to add is that I wouldn't recommend using a box or cylinder shape because these get stuck behind little things on the terrain more easily. Round shapes slide better.

**Reply by PhoenixIllusion** [reply]
Yes, those shapes were purely calling out and demonstrating the elevations, if testing to confirm explicit height restrictions. The flat-bottom shapes do make the height behavior explicit, but it is a poor user experience.

If you want to use a capsule to have a more natural flow of walking around, but also need an exact height for given steps, two options are to to either experiment, or to use geometry to calculate the capsule curvature values for a given capsule size. In both case you should produce an extra value to subtract from step heights, so that attempts to walk stairs will succeed with a given specific height and fail with anything more than that. 

The relationship of capsule radius to vertical offset should be fairly straightforward, and tied only to capsule radius. I believe it will need to be a Y value on the half-circle, such that the normal of the curvature is under the maxSlopeAngle of that particular character virtual, as decided by the `IsSlopeTooSteep` call in the Walk routine.

**Reply by jrouwe** [reply]
There is one more important parameter: `CharacterVirtualSettings.mSupportingVolume`. This plane defines up to which point the character can be supported. Anything above the plane will also result in the player sliding off. This was created so that you can't stand like this:

![image](https://github.com/user-attachments/assets/7137cd6d-1a91-49bd-b40a-41d8330571a1)

**Comment by LeXXik** [comment]
Yep, yep. Thank you, both! The curvature seems to have been the case. Switching to some "edgy" shape, like a box, shows the step height is respected. I was assuming the step height would be counted from the supporting volume, ignoring the character shape. That probably threw me off a bit.

Still not sure why the example doesn't use the step up/down values. In my app the step down and extra are not used. Made a little example on codesandbox. Just copy pasted the original example, increased the steps height a bit, but gave larger values for step up/down. Still only the first stairs can be used and going down from higher steps do not stick to lower steps, just natural gravity fall.

Pick preview 5000 for demo:
https://codesandbox.io/p/devbox/kind-rui-ygl3sq

**Reply by LeXXik** [reply]
Ah. it needs to be a negative number, as in opposite to up vector. My mistake. It works correctly after I changed it.
